modules/activiti-engine/src/main/java/org/activiti/engine/impl/persistence/entity/SignalEventSessionManager.java

```diff
@@ -0,0 +1,123 @@
+package org.activiti.engine.impl.persistence.entity;
+
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.activiti.engine.history.HistoricActivityInstance;
+import org.activiti.engine.impl.HistoricActivityInstanceQueryImpl;
+import org.activiti.engine.impl.bpmn.behavior.IntermediateThrowSignalEventActivityBehavior;
+import org.activiti.engine.impl.context.Context;
+import org.activiti.engine.impl.interceptor.CommandContext;
+import org.activiti.engine.impl.persistence.AbstractManager;
+import org.activiti.engine.impl.pvm.delegate.ActivityBehavior;
+import org.activiti.engine.impl.pvm.delegate.ActivityExecution;
+
+/**
+ * Keep track of fired events for process instance execution 
+ * in the current command context session and historic activity instance 
+ *
+ */
+public class SignalEventSessionManager extends AbstractManager {
+
+  /** collect fired events in process instance scope in the current command context session */
+  protected Map<String,Set<String>> firedSignalEventsSesssionRegistry = new HashMap<String, Set<String>>();
+  
+  /**
+   * Registers thrown signal events for execution in current session 
+   * 
+   * @param execution current execution
+   * @param eventName name of the event
+   */
+  public void registerThrowSignalEventByExecution(ActivityExecution execution, String eventName) {
+    String processInstanceId = execution.getProcessInstanceId();
+    
+    synchronized (firedSignalEventsSesssionRegistry) {

```

Comment:
Is the object to be locked too wide?
It is better to lock only each process instance.

In the current implementation, all signal throw processing is synchronized.

Reply:
@daisuke-yoshimoto  Each transaction will create a separate instance of SignalEventSessionManager. Throw signal event registration will be synchronized in the working memory of a single process instance session just in case if there are any concurrent executions in the current transaction. It is already locked by process instance by design.

Reply:
@igdianov 
If SignalEventSessionManager's instance is  separate in one transaction, you do not have to synchronize in the first place? Only one thread works per transaction, is not it?

Reply:
@daisuke-yoshimoto I put the synchronization block in `registerThrowSignalEventByExecution`  because I have noticed that execution may spawn async executions in separate threads in the same transaction.

Reply:
@igdianov 
> If it is always single thread per transaction

Yeah. I recognize that it is such behavior.

### activiti:async="true" && activiti:exclusive="true"

* Each parallel execution executes in serial and separate transactions.
* Although transactions are separate, control of in-memory Map is valid because threads are used the same.

### activiti:async="true" && activiti:exclusive="false"

* Each of the parallel executions is executed in separate transactions and separate threads.
* In the case of a cluster environment , since the execution of each execution is not executed by the same Activiti Engine, in-memory map control should not work.

For details, please check this document.
https://www.activiti.org/userguide/#bpmnConcurrencyAndTransactions